sstables/compaction_manager.cc

```diff
@@ -731,15 +731,17 @@ future<> compaction_manager::perform_cleanup(column_family* cf) {
         return make_exception_future<>(std::runtime_error(format("cleanup request failed: there is an ongoing cleanup on {}.{}",
             cf->schema()->ks_name(), cf->schema()->cf_name())));
     }
-    return rewrite_sstables(cf, sstables::compaction_options::make_cleanup(), [this] (const table& table) {
-        auto schema = table.schema();
+    return seastar::async([this, cf] {
+        auto schema = cf->schema();
         auto owned_ranges = service::get_local_storage_service().get_local_ranges(schema->ks_name());
         auto sstables = std::vector<sstables::shared_sstable>{};
-        const auto candidates = table.candidates_for_compaction();
+        const auto candidates = cf->candidates_for_compaction();
         std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&owned_ranges, schema] (const sstables::shared_sstable& sst) {
+            seastar::thread::maybe_yield();
             return owned_ranges.empty() || needs_cleanup(sst, owned_ranges, schema);
```

Comment:
If owned_ranges is sorted (and if not, we can sort it), then needs_cleanup can be optimized to take advantage of it.

Reply:
Yes, owned_ranges is sorted, so we can use that to optimize. However, compaction_manager::perform_cleanup is called on the slow path. It is more important to avoid stall.

Reply:
If it takes 16 seconds on some test it might take 16 minutes on a large cluster (if there is quadratic behavior). When something is this bad, we need to understand it, not just add yields.

Reply:
I think we know why it is slow: O(number_of_sstables * number_tokens_in_the_cluster)

Reply:
What are the numbers? Even if you have 100 sstables and 2000 tokens (8-node cluster), that's just 200,000 operations. Even if each operation takes 1usec (a huge amount) it should take 200ms.

So I think we don't know everything.

Reply:
I measured 2 nodes each with 2560 tokens on my dev machine.

```
diff --git a/sstables/compaction_manager.cc b/sstables/compaction_manager.cc
index b26a97fa7..20133c194 100644
--- a/sstables/compaction_manager.cc
+++ b/sstables/compaction_manager.cc
@@ -733,13 +733,20 @@ future<> compaction_manager::perform_cleanup(column_family* cf) {
     }
     return seastar::async([this, cf] {
         auto schema = cf->schema();
+        auto t1 = std::chrono::high_resolution_clock::now();
         auto owned_ranges = service::get_local_storage_service().get_local_ranges(schema->ks_name());
+        auto t2 = std::chrono::high_resolution_clock::now();
         auto sstables = std::vector<sstables::shared_sstable>{};
         const auto candidates = cf->candidates_for_compaction();
+        auto sz = candidates.size();
         std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&owned_ranges, schema] (const sstables::shared_sstable& sst) {
             seastar::thread::maybe_yield();
             return owned_ranges.empty() || needs_cleanup(sst, owned_ranges, schema);
         });
+        auto t3 = std::chrono::high_resolution_clock::now();
+        auto d1 = std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count();
+        auto d2 = std::chrono::duration_cast<std::chrono::nanoseconds>(t3 - t2).count();
+        cmlog.info("compaction_manager::perform_cleanup d1(get_ranges)={}, d2(needs_cleanup)={}, candidates={}", d1, d2, sz);
         rewrite_sstables(cf, sstables::compaction_options::make_cleanup(),
                 [sstables = std::move(sstables)] (const table&) { return sstables; }).get();
     });
```

INFO  2020-06-23 14:57:51,408 [shard 0] compaction_manager - compaction_manager::perform_cleanup ks=system_distributed, table=view_build_status, d1(get_ranges)=1204858379ns, d2(needs_cleanup)=2374ns, candidates=0


1.2s for get_ranges



INFO  2020-06-23 15:17:33,874 [shard 0] compaction_manager - compaction_manager::perform_cleanup ks=system_distributed, table=cdc_streams, d1(get_ranges)=5348718ns, d2(needs_cleanup)=5675718ns, candidates=1

5ms for get_ranges

5ms for 1 sstable -> 1000 sstables = 5s

Reply:
I guess you mean 5ms for needs_cleanup().

get_local_ranges() looks very slow. It is quadratic in terms of the number of tokens, and calculate_natural_endpoints() is itself slow. It is all pointless calculations, since the tokens change very rarely.

needs_cleanup() can be sped up significantly if owned_ranges was sorted.

Reply:
Yes 5ms for 1 sstable to run needs_cleanup.

Reply:
Yes 5ms for 1 sstable to run needs_cleanup.

> On Mon, Jun 22, 2020, 11:03 Asias He ***@***.***> wrote: @asias <https://github.com/asias> please see Avi's comments, and also please change target branch to next I know we merge PR into next branch. But aren't developers supposed to develop against master branch not next branch? Also when one sends a url for a patch set, the branch should be based on master and not next. Why it is important to change the target branch to next?
> @asias we develop against master but merge to next. The reason the conventional target branch for PRs is `next` is for merging using the github UI. It can also detect merge conflicts and warn about them and maybe in the future it could run some preliminary integration tests before merging.

Can we change the default to next for the sake of UI merge?

Reply:
We can't change the default, because then clones/forks will default to next, which is unstable.

Reply:
Note that cf352e7c1465e2ff16bb904eb0a74277b01312bc makes needs_cleanup() much faster.

Reply:
Yes. I chatted this with Raphael.

Reply:
> We can't change the default, because then clones/forks will default to next, which is unstable.

I meant changing the default branch for PR only. Sounds like github does not support it.